x.ps1

```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env pwsh
+
+# See x.sh for why these scripts exist.
+
+$xpy = Join-Path $PSScriptRoot x.py
+$xpy_args = @($xpy) + $args
+
+foreach ($python in "py", "python3", "python", "python2") {
+    # NOTE: this only tests that the command exists in PATH, not that it's actually
+    # executable. The latter is not possible in a portable way, see
+    # https://github.com/PowerShell/PowerShell/issues/12625.
+    if (Get-Command $python -ErrorAction SilentlyContinue) {
+        if ($python -eq "py") {
+            # Use python3, not python2
+            $xpy_args = @("-3") + $xpy_args
```

Comment:
Drive by comment as the person who uses PowerShell outside Windows:

This works fine on Windows but if someone (like me for example) prefers `py` over `python` or `python3` and create a symlink to emulate the one on Windows, this would fail as is (and `x.py` does today) because `-3` doesn't work on UNIX platforms:
```
./x.py test

unknown option: -3
usage: /Users/fox/.bin/py [option] ... [-c cmd | -m mod | file | -] [arg] ...
```

To improve the status quota, could you consider checking if its actually running on Windows first? This should do it and work with PS5 and PS6/7 (core).

```powershell
# Is this running in PowerShell Core?
if (Test-Path variable:IsWindows) {
    if ($IsWindows) {
        # Yes, and its on Windows
        # Use python3, not python2
        $xpy_args = @("-3") + $xpy_args
    } # if not, no extra arguments needed, someone made `py` on their own
} else {
    # Not PowerShell Core, must be Windows
     $xpy_args = @("-3") + $xpy_args
}
```

Reply:
This feels like too much of an edge case to support without risking it breaking all the time. IOW, I think it's important that we keep the number of configurations that these scripts need to support down as much as possible, in order to avoid future issues like what we had with `x.py`[^1], not to mention the benefit of simplifying maintenance/review/testing...

[^1]: Which definitely was made worse by the fact that it had tried to support many different systems under many different configurations auto-magically.

Anyway, even if you use powershell on unix, `./x.sh` should work fine and have no drawbacks, and I don't think there's an upside to using `x.ps1` in that case either.

So, if a check for non-windows is added to x.ps1, IMO it should be before anything else, and should basically bail out somehow. Concretely, one of the following would be my suggestion:

1. Give a nice error message that says to use `x.sh` instead.
2. Just invoke `x.sh` on behalf of the user and exit.

Either seems reasonable, although I'd lean towards the first to avoid having to deal with 2nd ending uncovering some other weird edge case that needs to be supported in several years.

Reply:
@BlackHoleFox your configuration is not Unix specific and I don't think we should do feature detection in these scripts other than checking binaries exist. If you want to have an alias that conflicts with an existing command, you should use x.py directly.

> this already has a shebang, so maybe support for powershell on unix is intended?

Yes, I intended to support PowerShell on Unix, mostly just to make these easy to test across platforms. (That's also the reason I'm avoiding your "have both ./x and ./x.cmd" idea)

Reply:
Alternatively we could just cycle through all of the permutations, running `py`, `py -3`, etc. to find one that fits our needs

Reply:
> Does PowerShell distinguish "failed to start process" from "process ended with an error code"? The latter isn't correct because x.py will exit with an error if the build fails.

We don't have a way to distinguish "this failed because x.py failed" and "this failed because py is actually just a symlink to python". So we can only try one of py, py -2, or py -3, and still reliably detect whether the command exists. I am willing to try `py` instead, but I am not willing to try and parse runtime errors.

Reply:
I'm not saying to parse runtime errors, just run the command with `--version` and see if it's python 3

Reply:
Won't `py --version` return the same thing?

Reply:
Again, the whole point of these scripts is that we *cannot* support all use cases in a single script and so we aren't going to try. If people have such a weird setup that `py` means something different on different platforms, they should be using x.py directly.